Added import test to fix import for parse_rc

icnocop opened the pull request:
Fixes #92

iafan commented on lib/Serge/Engine/Plugin/parse_rc.pm:
```diff
@@ -86,7 +86,7 @@ sub parse {
             my $str = $orig_str;
             $str =~ s/""/"/g;
-            $translated_str = &$callbackref($str, undef, $hint, undef, $lang);
+            $translated_str = &$callbackref($str, undef, $hint, undef, $lang, $hint);
```
This is not a proper fix. The last parameter is supposed to be a unique key (if available in the file), whereas you just use hint as a key there (which can be constructed in the parser in multiple ways).

I suggest moving this change out of the importer test PR.

iafan commented on t/engine.t:
```diff
@@ -9,7 +9,7 @@ use strict;
 # comma-separated list.  The following two examples are equivalent:
 #
-# perl t/enigne.t parse_json parse_strings
+# perl t/engine.t parse_json parse_strings
```
Can you externalize all typos / var name refactoring in a smaller PR that we can merge separately and which won't distract from the main importer PR?

iafan commented on t/engine.t:
```diff
@@ -42,7 +42,8 @@ sub Serge::Engine::file_mtime {
 
-my $thisdir = dirname(abs_path(__FILE__));
+my $this_dir = dirname(abs_path(__FILE__));
+my $tests_dir = catfile($this_dir, 'data', 'engine');
```
See my comment above for var name refactoring.

icnocop commented:
I've updated the PR with your code comments and added https://github.com/evernote/serge/pull/98 for the typo and refactoring as suggested.

As you can see, when I took out passing `$hint` as the last parameter to `$callbackref`, the test fails with the error:
```
Parser plugin didn't provide a key value in a callback. Importing translations with this plugin is not possible, unless you use `--disambiguate-keys` option.
```

Is that expected given the input file [application.rc](https://github.com/evernote/serge/blob/dbd2f3903088341ac95915e988ad53845cbdadb5/t/data/importer/parse_rc/00/resources/application.rc)?

If so, can you explain why, please?

All the keys in the `STRINGTABLE` are unique, so I don't expect any disambiguity.

Thank you.

icnocop commented:
When I took out passing `$hint` as the last parameter to `$callbackref`, the test fails with the error:
```
Parser plugin didn't provide a key value in a callback. Importing translations with this plugin is not possible, unless you use `--disambiguate-keys` option.
``

Is that expected given the input file [application.rc](https://github.com/evernote/serge/blob/dbd2f3903088341ac95915e988ad53845cbdadb5/t/data/importer/parse_rc/00/resources/application.rc)?

If so, can you explain why, please?

All the keys in the `STRINGTABLE` are unique, so I don't expect any disambiguity.

When I pass `$idstr` as the last parameter to `$callbackref`, the test fails with errors like:
```
Duplicate key 'BEGIN' found in source file
```

iafan commented on lib/Serge/Engine/Plugin/parse_rc.pm:
```diff
@@ -86,7 +86,7 @@ sub parse {
             my $str = $orig_str;
             $str =~ s/""/"/g;
-            $translated_str = &$callbackref($str, undef, $hint, undef, $lang);
+            $translated_str = &$callbackref($str, undef, $hint, undef, $lang, $idstr);
```
Now, this is a proper fix, and I'll be happy to merge the PR with this single-line change.

As for the importer, right now it is just a copy-paste of t/engine.t, and requires writing some additional tests; it also doesn't seem to do the actual import as the reference translations table in your test is empty. Would be more interesting to somehow hook the importing mode right into `engine.t`, where it would first generate the translations (and check them against reference files, and then use importer to import translations back and see that the reference import database matches the generated one.

icnocop replied:
Hi @iafan.

I updated the PR with another commit to fix more issues with import and parse_rc.
The translations table in the reference output are now populated correctly.

I agree that we should try to re-use some of the code in engine.t to consolidate the duplicate code this PR creates.
I'm not sure how to implement it as you suggested yet, but I'll try to at least get one step closer.

I'm wondering if you want me to create a new PR with just the changes in `Importer.pm` and `parse_rc.pm`?

Thank you.

icnocop replied:
In regards to "use importer to import translations back and see that the reference import database matches the generated one.", do you expect there to be a new `reference-output` database to check against the import? For example, `.\t\data\engine\parse_rc\00\reference-output\imported-database`? The `properties` and `translations` tables seem to be the only ones that are different.
